### PR TITLE
feat: eureka注册发现支持命名空间隔离

### LINKED ISSUE #1075: 通过eureka协议使用北极星支持命名空间隔离
**What is the feature you want to add?**

**Why do you want to add this feature?**

**How to implement this feature?**

**Additional context**

### PR DESCRIPTION
**Pull request by self-made-boy**
**Please provide issue(s) of this PR:**
Fixes #1075

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [x] ApiServer
- [ ] Auth
- [ ] Configuration
- [ ] Naming
- [ ] HealthCheck
- [ ] Metrics
- [ ] Docs
- [ ] Installation
- [ ] Performance and Scalability
- [ ] Test and Release

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any user-facing changes. This may include API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: apiserver/eurekaserver/access.go [hunk 1]
```diff
@@ -31,11 +31,12 @@ import (
...
-	ParamInstId string = "instId"
-	ParamValue  string = "value"
-	ParamVip    string = "vipAddress"
-	ParamSVip   string = "svipAddress"
+	ParamAppId      string = "appId"
+	ParamInstId     string = "instId"
+	ParamValue      string = "value"
+	ParamVip        string = "vipAddress"
+	ParamSVip       string = "svipAddress"
+	HeaderNamespace string = "X-Polaris-Namespace"
```
Comment (andrewshan) [inline apiserver/eurekaserver/access.go:39]: x-namespace

File: apiserver/eurekaserver/delta_worker.go [hunk 2]
```diff
@@ -45,6 +45,53 @@ func (l *Lease) Expired(curTimeSec int64, deltaExpireInterval time.Duration) boo
...
+		deltaExpireInterval:    deltaExpireInterval,
+		enableSelfPreservation: enableSelfPreservation,
+		namingServer:           namingServer,
+		healthCheckServer:      healthCheckServer,
+		workers:                workers,
+	}
+}
+
+func (a *ApplicationsWorkers) Get(namespace string) *ApplicationsWorker {
+	if work, exist := a.workers.Load(namespace); exist {
```
Comment (andrewshan) [inline apiserver/eurekaserver/delta_worker.go:78]: 这里最好加个锁

File: apiserver/eurekaserver/delta_worker.go [hunk 3]
```diff
@@ -45,6 +45,53 @@ func (l *Lease) Expired(curTimeSec int64, deltaExpireInterval time.Duration) boo
 	return curTimeSec-l.lastUpdateTimeSec >= deltaExpireInterval.Milliseconds()/1000
 }
 
+type ApplicationsWorkers struct {
+	interval               time.Duration
+	deltaExpireInterval    time.Duration
+	enableSelfPreservation bool
+	namingServer           service.DiscoverServer
+	healthCheckServer      *healthcheck.Server
+	workers                *sync.Map
```
Comment (andrewshan) [inline apiserver/eurekaserver/delta_worker.go:54]: 不用syncmap，用普通map+读写锁吧

File: apiserver/eurekaserver/model.go [hunk 4]
```diff
@@ -399,6 +399,7 @@ type StringMap map[string]interface{}
 
 // ReplicationInstance request for instance replicate
 type ReplicationInstance struct {
+	Namespace          string        `json:"namespace"`
```
Comment (andrewshan) [inline apiserver/eurekaserver/model.go:402]: 这个是没有的，eureka同步的时候，不会带这个过来

**Review by andrewshan — CHANGES_REQUESTED** [review]
1. 修改header的key
2. 使用读写锁+map
3. replicate的namespace不应该从body里面取
4. 这边发送replicate请求时，也应该把namespace带上

File: apiserver/eurekaserver/delta_worker.go [hunk 5]
```diff
@@ -45,6 +45,60 @@ func (l *Lease) Expired(curTimeSec int64, deltaExpireInterval time.Duration) boo
...
+func (a *ApplicationsWorkers) Get(namespace string) *ApplicationsWorker {
+	a.rwMutex.RLock()
+	work, exist := a.workers[namespace]
+	a.rwMutex.RUnlock()
+	if exist {
+		return work
+	} else {
+		a.rwMutex.Lock()
+		defer a.rwMutex.Unlock()
+		work := NewApplicationsWorker(a.interval, a.deltaExpireInterval, a.enableSelfPreservation,
```
Comment (andrewshan) [inline apiserver/eurekaserver/delta_worker.go:88]: 这里要再检查一下是否存在，在锁里面再进行一次判断
work, exist := a.workers[namespace]
if exist {
		return work
}

File: apiserver/eurekaserver/delta_worker.go [hunk 6]
```diff
@@ -45,6 +45,60 @@ func (l *Lease) Expired(curTimeSec int64, deltaExpireInterval time.Duration) boo
...
+	}
+}
+
+func (a *ApplicationsWorkers) Get(namespace string) *ApplicationsWorker {
+	a.rwMutex.RLock()
+	work, exist := a.workers[namespace]
+	a.rwMutex.RUnlock()
+	if exist {
+		return work
+	} else {
```
Comment (andrewshan) [inline apiserver/eurekaserver/delta_worker.go:85]: 这里没有必要写else

File: apiserver/eurekaserver/server.go [hunk 7]
```diff
@@ -265,6 +254,56 @@ func (h *EurekaServer) Initialize(ctx context.Context, option map[string]interfa
 	return nil
 }
 
+func parsePeersToReplicate(defaultNamespace string, replicatePeerObjs []interface{}) map[string][]string {
```
Comment (andrewshan) [inline apiserver/eurekaserver/server.go:257]: 这个方法要用测试用例来验证下

File: apiserver/eurekaserver/access.go [hunk 8]
```diff
@@ -271,13 +275,14 @@ func writeResponse(acceptValues map[string]bool, appsRespCache *ApplicationsResp
 
 // GetDeltaApplications 增量拉取服务实例信息
 func (h *EurekaServer) GetDeltaApplications(req *restful.Request, rsp *restful.Response) {
-	appsRespCache := h.worker.GetDeltaApps()
+	namespace := readNamespaceFromRequest(req, h.namespace)
+	appsRespCache := h.workers.Get(namespace).GetDeltaApps()
```
Comment (andrewshan) [inline apiserver/eurekaserver/access.go:279]: h.workers.Get(namespace)，提取成一个变量

**Review by chuntaojun — APPROVED** [review]

**Pull request merged**